cnd/src/storage.rs

```diff
@@ -158,197 +158,228 @@ impl Load<SwapContext> for Storage {
-        use crate::db::schema::swaps;
-
-        let (swap, halbit, herc20, secret_hash) = self
-            .db
-            .do_in_transaction::<_, _, anyhow::Error>(move |conn| {
-                let key = Text(id);
-
-                let swap: Swap = swaps::table
-                    .filter(swaps::local_swap_id.eq(key))
-                    .first(conn)?;
-
-                let halbit = Halbit::belonging_to(&swap).first::<Halbit>(conn)?;
-                let herc20 = Herc20::belonging_to(&swap).first::<Herc20>(conn)?;
-                let secret_hash = db::SecretHash::belonging_to(&swap)
-                    .first::<db::SecretHash>(conn)
-                    .optional()?;
-
-                Ok((swap, halbit, herc20, secret_hash))
-            })
-            .await
-            .context(db::Error::SwapNotFound)?;
+impl<A, B, TParamsA, TParamsB> Load<spawn::Swap<TParamsA, TParamsB>> for Storage
+where
+    Storage: LoadTables<A, B>,
+    TParamsA: IntoParams<ProtocolTable = A> + 'static,
+    TParamsB: IntoParams<ProtocolTable = B> + 'static,
+    A: 'static,
+    B: 'static,
+{
+    async fn load(&self, id: LocalSwapId) -> anyhow::Result<spawn::Swap<TParamsA, TParamsB>> {
+        let tab = self.load_tables(id).await?;
+        let role = tab.swap.role.0;
+        let secret_hash = derive_or_unwrap_secret_hash(id, self.seed, role, tab.secret_hash)?;
 
-        let role = swap.role.0;
-        let secret_hash = match role {
-            Role::Alice => {
-                let swap_seed = self.seed.derive_swap_seed(id);
-                SecretHash::new(swap_seed.derive_secret())
-            }
-            Role::Bob => secret_hash.ok_or_else(|| NoSecretHash(id))?.secret_hash.0,
-        };
+        let alpha = TParamsA::into_params(tab.alpha, id, self.seed, role, secret_hash)?;
+        let beta = TParamsB::into_params(tab.beta, id, self.seed, role, secret_hash)?;
 
-        let swap = spawn::Swap {
+        Ok(spawn::Swap {
             role,
-            alpha: build_herc20_params(herc20, secret_hash, id)?,
-            beta: build_halbit_params(halbit, secret_hash, id)?,
-            start_of_swap: swap.start_of_swap,
-        };
-
-        Ok(swap)
+            alpha,
+            beta,
+            start_of_swap: tab.swap.start_of_swap,
+        })
     }
 }
 
+/// Load data from tables, A and B are protocol tables.
 #[async_trait::async_trait]
-impl Load<spawn::Swap<halbit::Params, herc20::Params>> for Storage {
-    async fn load(
-        &self,
-        id: LocalSwapId,
-    ) -> anyhow::Result<spawn::Swap<halbit::Params, herc20::Params>> {
-        use crate::db::schema::swaps;
-
-        let (swap, halbit, herc20, secret_hash) = self
-            .db
-            .do_in_transaction::<_, _, anyhow::Error>(move |conn| {
-                let key = Text(id);
-
-                let swap: Swap = swaps::table
-                    .filter(swaps::local_swap_id.eq(key))
-                    .first(conn)?;
+pub trait LoadTables<A, B> {
```

Comment:
Nit: Rename to `LoadProtocolTables`?

Reply:
Yep can do.

Reply:
On second thoughts, this does not lead _just_ the protocol tables, it also includes the Swaps table and the SecretHash table. Admittedly this is a very generic name and will likely clash next time we need to load a set of different tables. Leaving as is for now.